### PR TITLE
Feature/tagging progress

### PR DESCRIPTION
**Pull request by alexanderbarnhill**

### CODE REVIEW

File: backend/service/tagging.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        time_index = 0
+        iteration_time = None
+        idx = 0
+        for lemmatized_keyword, (keyword, parents) in zip(
+            lemmatized_keywords, self.keywords.items()
+        ):
+            if self.cancelled:
+                break
+            if idx == 0:
+                progress_step = 80 / len(lemmatized_keywords)
```
Comment (Dimfred) [inline backend/service/tagging.py:60]: magic number

also cant the initialization of `progress_step` be moved out of the loop?

File: backend/service/tagging.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                time_index = 1
+
+            remaining_iterations = len(lemmatized_keywords) - idx
+            idx += 1
+            if iteration_time != - 1:
+                self.time_remaining = iteration_time * remaining_iterations
+            self.progress += progress_step
+        changed_docs = changed_docs.values()
+        self.status = 'TAGGING_JOB.DOC_UPDATE'
+        self.solr_docs.update(*changed_docs)
```
Comment (Dimfred) [inline backend/service/tagging.py:95]: add cancellation check before that

File: backend/service/tagging.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        self.solr_docs.update(*docs)
+
+    def stop(self):
+        self.cancelled = True
+
+
+class TaggingService:
+    jobs = {}
+
+    def __init__(self):
```
Comment (Dimfred) [inline backend/service/tagging.py:158]: unnecessary?

File: backend/service/tagging.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        iteration_time = None
+        idx = 0
+        for lemmatized_keyword, (keyword, parents) in zip(
+            lemmatized_keywords, self.keywords.items()
+        ):
+            if self.cancelled:
+                break
+            if idx == 0:
+                progress_step = 80 / len(lemmatized_keywords)
+            query = self.solr_docs.build_kwm_query(id_query, lemmatized_keyword)
```
Comment (Dimfred) [inline backend/service/tagging.py:61]: `build_kwm_query` should be part of the tagging job I think

File: backend/solr/docstorage.py [hunk 5]
```diff
@@ -136,25 +136,26 @@ def wipe_keywords(self):
             doc.keywords = []
             self.update(doc)
 
-    def apply_kwm(self, keywords: dict, *doc_ids: str) -> None:
+    def apply_kwm(self, keywords: dict, *doc_ids: str,) -> None:
```
Comment (Dimfred) [inline backend/solr/docstorage.py:139]: you can remove everything from here / move it to the tagging job

File: backend/service/tagging.py (hunk 4 above)
Comment (alexanderbarnhill) [reply backend/service/tagging.py:61]: It is?

Comment (Dimfred) [reply backend/service/tagging.py:61]: it's part of `SolrDocStorage` rightnow

Comment (Dimfred) [reply backend/service/tagging.py:61]: Or no nvm let it as is. Thats okay.

Comment (alexanderbarnhill) [reply backend/service/tagging.py:61]: Easy enough.

**Comment by Dimfred** [comment]
LGTM!

**Comment by Dimfred** [comment]
LGTM!

**Pull request closed without merging**